timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/ExclusiveLock.java

```diff
@@ -108,4 +95,35 @@ private void checkIsCurrentHolder(UUID requestId) {
         }
     }
 
+    @NotThreadSafe
+    private static class LockRequestQueue {
+
+        @SuppressWarnings("checkstyle:illegaltype")
+        private final LinkedHashMap<UUID, LockRequest> queue = Maps.newLinkedHashMap();
+
+        public void enqueue(LockRequest request) {
+            LockRequest existingRequest = queue.put(request.requestId, request);
+            if (existingRequest != null) {
+                queue.put(request.requestId, existingRequest);
+                throw new LoggableIllegalStateException(
+                        "Cannot enqueue the same request id twice.",
+                        SafeArg.of("requestId", request.requestId));
+            }
```

Comment:
from the docs on LinkedHashMap:
```
This linked list defines the iteration ordering,
 * which is normally the order in which keys were inserted into the map
 * (<i>insertion-order</i>).  Note that insertion order is not affected
 * if a key is <i>re-inserted</i> into the map.
```

But yeah, this exception being thrown is indicative of a product bug, it should never happen. It's arguably over-engineering to even put the old one back in